File: src/main/java/com/lessmarkup/engine/security/UserCacheImpl.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.security;

import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.dataobjects.User;
import com.lessmarkup.dataobjects.UserGroupMembership;
import com.lessmarkup.interfaces.cache.AbstractCacheHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.module.Implements;
import com.lessmarkup.interfaces.structure.CachedNodeInformation;
import com.lessmarkup.interfaces.structure.NodeAccessType;
import com.lessmarkup.interfaces.structure.NodeCache;
import com.lessmarkup.interfaces.structure.Tuple;
import com.lessmarkup.interfaces.system.UserCache;

import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.OptionalLong;

@Implements(UserCache.class)
public class UserCacheImpl extends AbstractCacheHandler implements UserCache {

    private OptionalLong userId;
    private final DomainModelProvider domainModelProvider;
    private final DataCache dataCache;
    
    private final List<Long> groups = new ArrayList<>();
    private boolean isRemoved;
    private boolean isAdministrator;
    private boolean isApproved;
    private boolean emailConfirmed;
    private String email;
    private String title;
    private boolean isBlocked;
    private OffsetDateTime unblockTime = null;
    private String properties;
    private OptionalLong avatarImageId;
    private OptionalLong userImageId;
    private String name;
    private final List<Tuple<CachedNodeInformation, NodeAccessType>> nodes = new ArrayList<>();

    @Inject
    public UserCacheImpl(DomainModelProvider domainModelProvider, DataCache dataCache) {
        super(new Class<?>[] { User.class });
        this.domainModelProvider = domainModelProvider;
        this.dataCache = dataCache;
    }
    
    private void initializeUser() {
        try (DomainModel domainModel = domainModelProvider.create()) {
            
            User user = domainModel.query().from(User.class).where(Constants.Data.ID_PROPERTY_NAME + " = $", userId.getAsLong()).firstOrDefault(User.class);

            if (user == null) {
                userId = OptionalLong.empty();
                return;
            }

            isAdministrator = user.isAdministrator();
            
            domainModel.query()
                    .from(UserGroupMembership.class)
                    .where("userId = $", userId.getAsLong())
                    .toList(UserGroupMembership.class, "UserGroupId")
                    .forEach(g -> { groups.add(g.getId()); });
            email = user.getEmail();
            title = user.getTitle();
            emailConfirmed = user.isEmailConfirmed();
            isBlocked = user.isBlocked();
            isApproved = user.isApproved();
            unblockTime = user.getUnblockTime();
            properties = user.getProperties();
            name = user.getName();
            avatarImageId = user.getAvatarImageId();
            userImageId = user.getUserImageId();

            if (isBlocked && unblockTime != null && unblockTime.isBefore(OffsetDateTime.now())) {
                isBlocked = false;
            }
        }
    }
    
    @Override
    public void initialize(OptionalLong objectId) {
        userId = objectId;
        
        if (userId.isPresent()) {
            initializeUser();
        }
        
        dataCache.get(NodeCache.class).getNodes().stream()
                .forEach(node -> { 
                    NodeAccessType type = node.checkRights(this, userId);
                    if (type == NodeAccessType.NO_ACCESS) {
                        return;
                    }
                    nodes.add(new Tuple<>(node, type));
                });
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public boolean isRemoved() {
        return isRemoved;
    }

    @Override
    public boolean isAdministrator() {
        return isAdministrator;
    }

    @Override
    public boolean isApproved() {
        return isApproved;
    }

    @Override
    public boolean isEmailConfirmed() {
        return emailConfirmed;
    }

    @Override
    public List<Long> getGroups() {
        return groups;
    }

    @Override
    public String getEmail() {
        return email;
    }

    @Override
    public String getTitle() {
        return title;
    }

    @Override
    public boolean isBlocked() {
        return isBlocked;
    }

    @Override
    public String getProperties() {
        return properties;
    }

    @Override
    public OptionalLong getAvatarImageId() {
        return avatarImageId;
    }

    @Override
    public OptionalLong getUserImageId() {
        return userImageId;
    }

    @Override
    public List<Tuple<CachedNodeInformation, NodeAccessType>> getNodes() {
        return nodes;
    }

    @Override
    public OffsetDateTime getUnblockTime() {
        return unblockTime;
    }
    
}

File: src/main/scala/com/lessmarkup/engine/security/UserCacheImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.security

import java.time.OffsetDateTime
import com.google.inject.Inject
import com.lessmarkup.Constants
import com.lessmarkup.dataobjects.{User, UserGroupMembership}
import com.lessmarkup.interfaces.annotations.{Implements, NodeAccessType, UseInstanceFactory}
import com.lessmarkup.interfaces.cache._
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.structure.{CachedNodeInformation, NodeCache}
import com.lessmarkup.interfaces.system.UserCache

class UserCacheFactory @Inject() (domainModelProvider: DomainModelProvider, dataCache: DataCache) extends InstanceFactory {
  override def createInstance(params: Any*): CacheHandler = {
    new UserCacheImpl(domainModelProvider, dataCache, params.head.asInstanceOf)
  }
}

@Implements(classOf[UserCache])
@UseInstanceFactory(classOf[UserCacheFactory])
class UserCacheImpl (domainModelProvider: DomainModelProvider, dataCache: DataCache, userId: Long) extends AbstractCacheHandler(Array[Class[_]](classOf[User])) with UserCache {

  private val user = loadUser

  private def loadUser: Option[User] = {
    val domainModel: DomainModel = domainModelProvider.create
    try {
      domainModel.query
        .from(classOf[User])
        .where(Constants.DataIdPropertyName + " = $", userId)
        .first(classOf[User], None)
    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  def getName: String = if (user.isDefined) user.get.name else ""

  def isRemoved: Boolean = user.isEmpty || user.get.removed

  def isAdministrator: Boolean = user.isDefined && user.get.administrator

  def isApproved: Boolean = user.isDefined && user.get.approved

  def isEmailConfirmed: Boolean = user.isDefined && user.get.emailConfirmed

  def getGroups: Seq[Long] = {

    if (user.isEmpty) {
      return List()
    }

    val domainModel: DomainModel = domainModelProvider.create
    try {
      domainModel.query
        .from(classOf[UserGroupMembership])
        .where("userId = $", userId).toList(classOf[UserGroupMembership], Option("UserGroupId")).map(_.userGroupId)
    } finally {
      domainModel.close()
    }
  }

  def getEmail: String = if (user.isDefined) user.get.email else ""

  def getTitle: String = if (user.isDefined) user.get.title.getOrElse("") else ""

  def isBlocked: Boolean = {

    if (user.isEmpty) {
      return true
    }

    if (!user.get.blocked) {
      return false
    }

    if (user.get.unblockTime.isDefined && user.get.unblockTime.get.isBefore(OffsetDateTime.now)) {
      return false
    }

    true
  }

  def getProperties: String = if (user.isDefined) user.get.properties.getOrElse("") else ""

  def getAvatarImageId: Option[Long] = {
    if (user.isDefined) {
      val imageId = user.get.avatarImageId
      if (imageId.isDefined) Option(imageId.get) else None
    } else {
      None
    }
  }

  def getUserImageId: Option[Long] = {
    if (user.isDefined) {
      val imageId = user.get.userImageId
      if (imageId.isDefined) Option(imageId.get) else None
    } else {
      None
    }
  }

  def getNodes: List[(CachedNodeInformation, NodeAccessType)] = {

    dataCache.get(classOf[NodeCache])
      .getNodes
      .toList
      .map(n => {
        val rights = n.checkRights(this, Option(userId))
        if (rights == NodeAccessType.NO_ACCESS) {
          None
        } else {
          Option((n, rights))
        }
      })
      .filter(_.isDefined)
      .map(_.get)
  }

  def getUnblockTime: Option[OffsetDateTime] = {
    if (user.isEmpty) {
      None
    } else {
      user.get.unblockTime
    }
  }
}